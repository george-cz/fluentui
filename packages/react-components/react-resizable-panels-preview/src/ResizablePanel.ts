export * from './components/ResizablePanel/index';
