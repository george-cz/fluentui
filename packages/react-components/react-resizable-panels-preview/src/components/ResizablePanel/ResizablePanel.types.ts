import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';

export type ResizablePanelSlots = {
  root: Slot<'div'>;
};

/**
 * ResizablePanel Props
 */
export type ResizablePanelProps = ComponentProps<ResizablePanelSlots> & {};

/**
 * State used in rendering ResizablePanel
 */
export type ResizablePanelState = ComponentState<ResizablePanelSlots>;
// TODO: Remove semicolon from previous line, uncomment next line, and provide union of props to pick from ResizablePanelProps.
// & Required<Pick<ResizablePanelProps, 'propName'>>
