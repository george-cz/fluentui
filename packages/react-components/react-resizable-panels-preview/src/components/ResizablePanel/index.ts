export * from './ResizablePanel';
export * from './ResizablePanel.types';
export * from './renderResizablePanel';
export * from './useResizablePanel';
export * from './useResizablePanelStyles.styles';
