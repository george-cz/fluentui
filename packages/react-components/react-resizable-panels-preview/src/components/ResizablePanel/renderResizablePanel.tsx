/** @jsxRuntime automatic */
/** @jsxImportSource @fluentui/react-jsx-runtime */

import { assertSlots } from '@fluentui/react-utilities';
import type { ResizablePanelState, ResizablePanelSlots } from './ResizablePanel.types';

/**
 * Render the final JSX of ResizablePanel
 */
export const renderResizablePanel_unstable = (state: ResizablePanelState) => {
  assertSlots<ResizablePanelSlots>(state);

  // TODO Add additional slots in the appropriate place
  return <state.root />;
};
