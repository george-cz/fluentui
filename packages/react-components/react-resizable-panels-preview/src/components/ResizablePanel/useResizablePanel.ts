import * as React from 'react';
import { getIntrinsicElementProps, slot } from '@fluentui/react-utilities';
import type { ResizablePanelProps, ResizablePanelState } from './ResizablePanel.types';

/**
 * Create the state required to render ResizablePanel.
 *
 * The returned state can be modified with hooks such as useResizablePanelStyles_unstable,
 * before being passed to renderResizablePanel_unstable.
 *
 * @param props - props from this instance of ResizablePanel
 * @param ref - reference to root HTMLDivElement of ResizablePanel
 */
export const useResizablePanel_unstable = (
  props: ResizablePanelProps,
  ref: React.Ref<HTMLDivElement>,
): ResizablePanelState => {
  return {
    // TODO add appropriate props/defaults
    components: {
      // TODO add each slot's element type or component
      root: 'div',
    },
    // TODO add appropriate slots, for example:
    // mySlot: resolveShorthand(props.mySlot),
    root: slot.always(
      getIntrinsicElementProps('div', {
        ref,
        ...props,
      }),
      { elementType: 'div' },
    ),
  };
};
