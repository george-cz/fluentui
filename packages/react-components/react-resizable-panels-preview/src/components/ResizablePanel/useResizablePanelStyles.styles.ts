import { makeStyles, mergeClasses } from '@griffel/react';
import type { SlotClassNames } from '@fluentui/react-utilities';
import type { ResizablePanelSlots, ResizablePanelState } from './ResizablePanel.types';

export const resizablePanelClassNames: SlotClassNames<ResizablePanelSlots> = {
  root: 'fui-ResizablePanel',
  // TODO: add class names for all slots on ResizablePanelSlots.
  // Should be of the form `<slotName>: 'fui-ResizablePanel__<slotName>`
};

/**
 * Styles for the root slot
 */
const useStyles = makeStyles({
  root: {
    // TODO Add default styles for the root element
  },

  // TODO add additional classes for different states and/or slots
});

/**
 * Apply styling to the ResizablePanel slots based on the state
 */
export const useResizablePanelStyles_unstable = (state: ResizablePanelState): ResizablePanelState => {
  const styles = useStyles();
  state.root.className = mergeClasses(resizablePanelClassNames.root, styles.root, state.root.className);

  // TODO Add class names to slots, for example:
  // state.mySlot.className = mergeClasses(styles.mySlot, state.mySlot.className);

  return state;
};
