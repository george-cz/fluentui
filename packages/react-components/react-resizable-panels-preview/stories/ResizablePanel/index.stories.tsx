import { ResizablePanel } from '@fluentui/react-resizable-panels-preview';

import descriptionMd from './ResizablePanelDescription.md';
import bestPracticesMd from './ResizablePanelBestPractices.md';

export { Default } from './ResizablePanelDefault.stories';

export default {
  title: 'Preview Components/ResizablePanel',
  component: ResizablePanel,
  parameters: {
    docs: {
      description: {
        component: [descriptionMd, bestPracticesMd].join('\n'),
      },
    },
  },
};
