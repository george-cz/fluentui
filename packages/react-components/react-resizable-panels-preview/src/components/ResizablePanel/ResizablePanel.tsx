import * as React from 'react';
import type { ForwardRefComponent } from '@fluentui/react-utilities';
import { useCustomStyleHook_unstable } from '@fluentui/react-shared-contexts';
import { useResizablePanel_unstable } from './useResizablePanel';
import { renderResizablePanel_unstable } from './renderResizablePanel';
import { useResizablePanelStyles_unstable } from './useResizablePanelStyles.styles';
import type { ResizablePanelProps } from './ResizablePanel.types';

/**
 * ResizablePanel component - TODO: add more docs
 */
export const ResizablePanel: ForwardRefComponent<ResizablePanelProps> = React.forwardRef((props, ref) => {
  const state = useResizablePanel_unstable(props, ref);

  useResizablePanelStyles_unstable(state);
  // TODO update types in packages/react-components/react-shared-contexts/src/CustomStyleHooksContext/CustomStyleHooksContext.ts
  // https://github.com/microsoft/fluentui/blob/master/rfcs/react-components/convergence/custom-styling.md
  useCustomStyleHook_unstable('useResizablePanelStyles_unstable')(state);
  return renderResizablePanel_unstable(state);
});

ResizablePanel.displayName = 'ResizablePanel';
