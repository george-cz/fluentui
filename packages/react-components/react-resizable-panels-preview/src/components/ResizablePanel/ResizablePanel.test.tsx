import * as React from 'react';
import { render } from '@testing-library/react';
import { isConformant } from '../../testing/isConformant';
import { ResizablePanel } from './ResizablePanel';

describe('ResizablePanel', () => {
  isConformant({
    Component: ResizablePanel,
    displayName: 'ResizablePanel',
  });

  // TODO add more tests here, and create visual regression tests in /apps/vr-tests

  it('renders a default state', () => {
    const result = render(<ResizablePanel>Default ResizablePanel</ResizablePanel>);
    expect(result.container).toMatchSnapshot();
  });
});
