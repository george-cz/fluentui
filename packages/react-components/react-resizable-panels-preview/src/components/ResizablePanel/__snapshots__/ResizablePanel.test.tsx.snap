// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResizablePanel renders a default state 1`] = `
<div>
  <div
    class="fui-ResizablePanel"
  >
    Default ResizablePanel
  </div>
</div>
`;
